piker/_cacheables.py

```diff
@@ -0,0 +1,222 @@
+
+        return wrapper
+
+    return decorator
+
+
+_cache: dict[str, 'Client'] = {}  # noqa
+
+
+# XXX: this mis mostly an alt-implementation of
+# maybe_open_ctx() below except it uses an async exit statck.
+# ideally wer pick one or the other.
+@asynccontextmanager
+async def open_cached_client(
+    brokername: str,
+) -> 'Client':  # noqa
+    '''Get a cached broker client from the current actor's local vars.
+
+    If one has not been setup do it and cache it.
+    '''
+    global _cache
+
+    clients = _cache.setdefault('clients', {'_lock': trio.Lock()})
+
+    # global cache task lock
+    lock = clients['_lock']
+
+    client = None
+
+    try:
+        log.info(f"Loading existing `{brokername}` client")
+
+        async with lock:
+            client = clients[brokername]
+            client._consumers += 1
+
+        yield client
+
+    except KeyError:
+        log.info(f"Creating new client for broker {brokername}")
+
+        async with lock:
+            brokermod = get_brokermod(brokername)
+            exit_stack = AsyncExitStack()
+
+            client = await exit_stack.enter_async_context(
+                brokermod.get_client()
+            )
+            client._consumers = 0
+            client._exit_stack = exit_stack
+            clients[brokername] = client
+
+        yield client
+
+    finally:
+        if client is not None:
+            # if no more consumers, teardown the client
+            client._consumers -= 1
+            if client._consumers <= 0:
+                await client._exit_stack.aclose()
+
+
+class cache:
+    '''Globally (processs wide) cached, task access to a
+    kept-alive-while-in-use async resource.
+
+    '''
+    lock = trio.Lock()
+    users: int = 0
+    values: dict[tuple[str, str], tuple[AsyncExitStack, Any]] = {}
+    nurseries: dict[int, Optional[trio.Nursery]] = {}
+    no_more_users: Optional[trio.Event] = None
+
+    @classmethod
+    async def run_ctx(
+        cls,
+        mng,
+        key,
+        task_status: TaskStatus[T] = trio.TASK_STATUS_IGNORED,
+
+    ) -> None:
+        async with mng as value:
+
+            cls.no_more_users = trio.Event()
+            cls.values[key] = value
+            task_status.started(value)
+            try:
+                await cls.no_more_users.wait()
+            finally:
+                value = cls.values.pop(key)
+                # discard nursery ref so it won't be re-used (an error)
+                cls.nurseries.pop(id(mng))
+
+
+@asynccontextmanager
+async def maybe_open_ctx(
+
+    key: Hashable,
+    mngr: AsyncContextManager[T],
+
+) -> (bool, T):
+    '''Maybe open a context manager if there is not already a cached
+    version for the provided ``key``. Return the cached instance on
+    a cache hit.
+
+    '''
+
+    await cache.lock.acquire()
+
+    ctx_key = id(mngr)
+
+    # TODO: does this need to be a tractor "root nursery"?
+    async with maybe_open_nursery(cache.nurseries.get(ctx_key)) as n:
```

Comment:
I'm still slightly unclear how the teardown part of lifetime works here; pretty sure it's going to tear down when the *first* consumer task is complete instead of when the *last* arriving consumer does.

In order to get the latter behavior we might need to have an *actor global* nursery that's brought up with the runtime / the consumer process?

Not sure this absolutely must be addressed right now since usually the creator task stays up as long as the app / daemon which is using the feed.